src/sqlparser/src/parser.rs

```diff
@@ -687,9 +687,29 @@ impl Parser {
                             Expr::Row(exprs)
                         }
                     };
+
                 self.expect_token(&Token::RParen)?;
                 if self.peek_token() == Token::Period && matches!(expr, Expr::Nested(_)) {
                     self.parse_struct_selection(expr)
+                } else if self.peek_token() == Token::Arrow && let Some(args) = {
+                    match expr.clone() {
+                        Expr::Row(args) => Some(args),
+                        Expr::Nested(arg) => Some(vec![*arg]),
+                        _ => None,
+                    }
+                } {
+                    let args: Vec<_> = args.into_iter().map(|arg| {
+                        let Expr::Identifier(name) = arg else {
+                            return parser_err!(format!(
+                                "Expected `Ident` as lambda function arg, found: {}", arg
+                            ));
+                        };
+
+                        Ok(name)
+                    }).try_collect()?;
+                    self.expect_token(&Token::Arrow)?;
+                    let body = self.parse_expr()?;
```

Comment:
```
dev=> select ('{"a":1}'::jsonb) -> 'a';
ERROR:  QueryError: sql parser error: Expected `Ident` as lambda function arg, found: CAST('{"a":1}' AS jsonb)

dev=> with t(v) as (select '{"a":1}'::jsonb) select (v) -> 'a' is null from t;
ERROR:  QueryError: Bind error: failed to bind expression: (v) -> 'a' IS NULL

Caused by:
  Feature is not yet implemented: JsonbAccessInner[Jsonb, Boolean]
Tracking issue: https://github.com/risingwavelabs/risingwave/issues/112
```

Both valid json access usages are mis-parsed, although the parentheses are redundant.

Or we may claim it is always unnecessary to use parentheses to the left of `->`, because all operators with lower precedence (`BETWEEN`/`<>`/`IS`/`NOT`/`AND`/`OR`) can only produce `boolean`. As a result:
* all valid expression can still be expressed by removing unnecessary parentheses
* invalid expression may have a cryptic error message
* pg expressions with unnecessary parentheses are rejected - minor incompatibility

Reply:
Looks good to me